<h1>Project: United Kingdom Road Accident Data Analysis</h1>

<h3>Inclusive Years:2019-2022</h3>

<h4>Analyst: Jamaica M. Loria</h4>

<h1>Importing Libraries / Data Preparation</h1>

In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
import numpy as np
from scipy.stats import f_oneway

In [2]:
accident = pd.read_csv('dataset\\accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Data Cleansing / Filling Up Missing Values in the Data Frame</h1>

In [4]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [5]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unaccounted')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')


In [6]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [7]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

<h1>Categorical Data Fields</h1>

In [8]:
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Latitude'] = accident['Latitude'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Longitude'] = accident['Longitude'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [9]:
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [10]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Extracting Information using pandas Date Time</h1>

In [11]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Month'] = accident['Accident Date'].dt.month
accident['Day'] = accident['Accident Date'].dt.day
accident['DayOfWeek'] = accident['Accident Date'].dt.dayofweek
accident['Daylight'] = accident['Light_Conditions'].str.contains('Daylight')

In [12]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
Daylight                   0
dtype: int64

<h1>INSIGHT 1</h1>

<h1>1.What Area has the Highest number of Accidents?</h1>

In [56]:
area_severity = accident['Urban_or_Rural_Area'].value_counts()
area_severity

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h3>Insight 1.1: Based on the result, 63.83% of accidents happened in Urban Area</h3>

<h1>INSIGHT 2</h1>

<h1>2.What Severity Level has the Highest Amount of Accidents?</h1>

In [57]:
severity_counts = accident['Accident_Severity'].value_counts()
severity_counts

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>Insight 2.1: Based on the result, 85.34% of accidents has Slight Severity</h3>

<h1>INSIGHT 3</h1>

<h1>3. How many incidents happened in an Urban or Rural Area in the Year 2019 </h1>

In [59]:
year_2019 = accident[accident['Year']== 2019]
year_2019['Urban_or_Rural_Area'].value_counts().sum()

182115

<h3>Insight 3.1: There is a total number of 182, 115 incidents that happend in a Urban or Rural Area in the year 2019</h3>

<h1>INSIGHT 4</h1>

<h1>4. What vehicle type causes the most and least accident?</h1>

In [16]:
accident.groupby('Vehicle_Type')['Number_of_Casualties'].sum().sort_values(ascending=False)

Vehicle_Type
Car                                      676692
Van / Goods 3.5 tonnes mgw or under       46271
Bus or coach (17 or more pass seats)      34915
Motorcycle over 500cc                     34879
Goods 7.5 tonnes mgw and over             23397
Motorcycle 125cc and under                20348
Taxi/Private hire car                     18195
Motorcycle over 125cc and up to 500cc     10286
Motorcycle 50cc and under                 10167
Goods over 3.5t. and under 7.5t            8308
Other vehicle                              7554
Minibus (8 - 16 passenger seats)           2659
Agricultural vehicle                       2613
Pedal cycle                                 270
Data missing or out of range                  9
Ridden horse                                  5
Name: Number_of_Casualties, dtype: int64

<h3>Insight 4.1: 102.42% of cars causes the most casualties while 0.0008% of ridden horse causes the least casualties</h3>

<h1>INSIGHT 5</h1>

<h1>5. How many Car accidents happened in a Fine no high winds weather conditions</h1>

In [17]:
accident[(accident['Vehicle_Type'] == 'Car') & (accident['Weather_Conditions'] == 'Fine no high winds')].info()

<class 'pandas.core.frame.DataFrame'>
Index: 392668 entries, 0 to 660676
Data columns (total 19 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    392668 non-null  category      
 1   Accident_Severity        392668 non-null  category      
 2   Accident Date            392668 non-null  datetime64[ns]
 3   Latitude                 392668 non-null  category      
 4   Light_Conditions         392668 non-null  category      
 5   District Area            392668 non-null  category      
 6   Longitude                392668 non-null  category      
 7   Number_of_Casualties     392668 non-null  int64         
 8   Number_of_Vehicles       392668 non-null  int64         
 9   Road_Surface_Conditions  392668 non-null  category      
 10  Road_Type                392668 non-null  category      
 11  Urban_or_Rural_Area      392668 non-null  category      
 12  Weather_Conditions   

<h3>Insight 5.1: There are approximately 59.43% car accident in a fine no high winds weather conditions</h3>

<h1>INSIGHT 6</h1>

<h1>6. What year has the most casualties by accidents?</h1>

In [18]:
accident.groupby('Year')['Number_of_Casualties'].sum().sort_values(ascending=False)

Year
2019    247780
2020    230905
2021    222146
2022    195737
Name: Number_of_Casualties, dtype: int64

<h3>Insight 6.1: Year 2019 has the most casualties with a total number of 247,780 accident</h3>

<h1>INSIGHT 7</h1>

<h1>7. What Month has the most casualties by accidents?</h1>

In [19]:
accident.groupby('Month')['Number_of_Casualties'].sum().sort_values(ascending=False)

Month
11    81453
10    80302
7     78490
5     76796
6     76296
9     76053
8     75280
3     73076
1     71038
12    70651
4     70620
2     66513
Name: Number_of_Casualties, dtype: int64

<h3>Insight 7.1: The month of November has the most casualties with a total number of 81,453 accident</h3>

<h1>INSIGHT 8</h1>

<h1>8. Is their any correlation between Number of Vehicles and the Number of Casualties?</h1>

In [20]:
insight_one = accident['Number_of_Vehicles'].corr(accident['Number_of_Casualties'])
insight_one

0.22888886126927557

<h3>Insight 8.1: No, there is no correlation between the total number of vehicles and number of casualties</h3>

<h1>INSIGHT 9</h1>

<h1>9. What District causes the most accidents?</h1>

In [21]:
accident.groupby('District Area')['Number_of_Casualties'].sum().sort_values(ascending=False)

District Area
Birmingham                   18674
Leeds                        12438
Manchester                    9437
Bradford                      9268
Liverpool                     8669
                             ...  
London Airport (Heathrow)      217
Clackmannshire                 204
Shetland Islands               202
Orkney Islands                 154
Clackmannanshire               115
Name: Number_of_Casualties, Length: 422, dtype: int64

<h3>Insight 9.1: The District Area of Birminghan has cost the most deaths</h3>

<h1>INSIGHT 10</h1>

<h1>10. How many casualties happen in Daylight?</h1>

In [22]:
accident.groupby('Daylight')['Number_of_Casualties'].sum().sort_values(ascending=False)

Daylight
True     649889
False    246679
Name: Number_of_Casualties, dtype: int64

<h3>Insight 10.1: Approximately 98.37% of the accidents happened in Daylight Condition </h3>

<h1>INSIGHT 11</h1>

<h1>11. How many vehicles cause each severity level</h1

In [23]:
accident.groupby('Accident_Severity')['Number_of_Vehicles'].count()

Accident_Severity
Fatal        8661
Serious     88217
Slight     563801
Name: Number_of_Vehicles, dtype: int64

<h3>Insight 11.1: There happens to be 1.31% vehicles that has caused fatal severity, 13.35% caused serious severity, and 85.34% caused slight severity</h3>

<h1>INSIGHT 12</h1>

<h1>12. What type of Road causes the most casualties</h1>

In [24]:
accident.groupby('Road_Type')['Number_of_Casualties'].count()

Road_Type
Dual carriageway       99424
One way street         13559
Roundabout             43992
Single carriageway    492143
Slip road               7041
unaccounted             4520
Name: Number_of_Casualties, dtype: int64

<h3>Insight 12.1: Single carriageway showed the highest number of accident with 74.49% of casualties </h3>

<h1>INSIGHT 13</h1>

<h1>13. Is there any correlation between number of vehicles and number of casualties</h1>

In [25]:
vehicle_casualties = accident ['Number_of_Casualties'].corr(accident['Number_of_Vehicles'])

In [26]:
vehicle_casualties

0.2288888612692756

<h3>Insight 13.1: There is no Correlation between number of vehicles and number of casualties</h3>

<h1>Insight 14</h1>

<h1>14. Is there any correlation between Latitude and Number of Casualties</h1>

In [27]:
latitude_casualties = accident ['Latitude'].corr(accident['Number_of_Casualties'])

In [28]:
latitude_casualties

0.03220129464996809

<h3>Insight 14.1: There is no Correlation between Latitude and Number of Casualties</h3>

<h1>INSIGHT 15</h1>

<h1>14. Is there any correlation between Longitude and Number of Casualties</h1>

In [60]:
longitude_casualties = accident ['Longitude'].corr(accident['Number_of_Casualties'])

In [61]:
longitude_casualties

-0.04040873919498823

<h3>Insight 15.1: There is no Correlation between Longitude and Number of Casualties</h3>

<h1>INSIGHT 16</h1>

<h1>16. What Road Surface Conditions does accident occur frequently?</h1>

In [62]:
accident['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     447821
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
unaccounted                726
Name: count, dtype: int64

<h3>Insight 16.1: Based on the result, 67.78% of accident occur frequently in a Dry Road Surface Conditions</h3>

<h1>INSIGHT 17</h1>

<h1>17. Is there any correlation between the Urban or Rural Area and Number of vehicles?</h1>

In [31]:
f_stats, p_value = f_oneway(accident[accident['Urban_or_Rural_Area'] == 'Urban']['Number_of_Vehicles'],
                            accident[accident['Urban_or_Rural_Area'] == 'Rural']['Number_of_Vehicles'])
print(p_value)

9.09607802972544e-198


<h3>Insight 17.1: Using the one way anova, analysis of variance, there is a correlation between the null value of Urban or Rural Area and Number of vehicles</h3>


<h1>INSIGHT 18</h1>

<h1>18. In what year has the most number of vehicles?</h1>

In [32]:
accident.groupby('Year')['Number_of_Vehicles'].sum().sort_values(ascending=False)

Year
2019    334966
2020    311604
2021    298687
2022    264615
Name: Number_of_Vehicles, dtype: int64

<h3>Insight 18.1: Year 2019 has the most number of vehicles with an estimation of 50.70% vehicles</h3>

<h1>INSIGHT 19</h1>

<h1>19. Is their any correlation between Number of Vehicles and in each category of Accident Severity? </h1>

In [33]:
slight_corr = accident[(accident['Accident_Severity'] == 'Slight')]['Number_of_Vehicles'].corr (accident[(accident['Accident_Severity'] == 'Slight')]['Number_of_Vehicles'])
print (f'Slight: {slight_corr}')

serious_corr = accident[(accident['Accident_Severity'] == 'Serious')]['Number_of_Vehicles'].corr (accident[(accident['Accident_Severity'] == 'Serious')]['Number_of_Vehicles'])
print (f'Serious: {serious_corr}')

fatal_corr = accident[(accident['Accident_Severity'] == 'Fatal')]['Number_of_Vehicles'].corr (accident[(accident['Accident_Severity'] == 'Fatal')]['Number_of_Vehicles'])
print (f'Fatal: {fatal_corr}')





Slight: 1.0
Serious: 1.0
Fatal: 1.0


<h3>19.1: Based on my findings from the dataset, I can infer that there is a correlation with regards to Accident Severity and Number of Vehicles in each category of severity.</h3>

<h1>INSIGHT 20</h1>

<h1>20. What District Area and it's Light Level causes the most casualties?</h1>

In [44]:
casualties_area_condition = np.round(accident.groupby(['District Area','Light_Conditions'])['Number_of_Casualties'].mean(),2)

casualties_area_condition.idxmax()

<h3>20.1: The Salisbury area and Darkness - lights unlit conditions causes the most casulaties</h3>

<h1>INSIGHT 21</h1>

<h1>21. What is the highest average number of deaths in each Light Conditions?</h1>

In [67]:
area = accident[accident['Weather_Conditions']== 'Fine + high winds']

In [79]:
area2 = np.round(area.groupby(['Light_Conditions'])['Number_of_Casualties'].mean(),2)
area2

Light_Conditions
Darkness - lighting unknown    1.39
Darkness - lights lit          1.37
Darkness - lights unlit        1.47
Darkness - no lighting         1.52
Daylight                       1.37
Name: Number_of_Casualties, dtype: float64

<h3>Insight 21.1: The Darkness - no lighting conditions has the highest number of deaths with an average of 1.52</h3>

<h1>INSIGHT 22</h1>